### PR TITLE
More PeeringState and related cleanups to ease use in crimson

### PR DESCRIPTION
**Pull request by athanatos**
Also switches all users in osd/ to use std::optional.

### CODE REVIEW

**Comment by athanatos** [comment]
Test results prior to rebase and valgrind fix: http://pulpito.ceph.com/sjust-2019-05-07_19:11:42-rados-sjust-wip-peering-state-cleanup-distro-basic-smithi/
Test rerun with valgrind fix: http://pulpito.ceph.com/sjust-2019-05-09_00:02:14-rados-sjust-wip-peering-state-cleanup-distro-basic-smithi/ (still has mgr valgrind errors which I think are unrelated)

**Comment by athanatos** [comment]
@neha-ojha The valgrind fix is first -- let me know if you'd rather I split it out.

**Comment by neha-ojha** [comment]
> @neha-ojha The valgrind fix is first -- let me know if you'd rather I split it out.

You already have the tracker reference in your commit so it's fine to leave it here.

File: src/osd/osd_types.h [hunk 1]
```diff
@@ -5925,6 +5926,27 @@ struct pool_pg_num_history_t {
 };
 WRITE_CLASS_ENCODER(pool_pg_num_history_t)
 
+// prefix pgmeta_oid keys with _ so that PGLog::read_log_and_missing() can
+// easily skip them
+static const string_view infover_key = "_infover"sv;
```
Comment (jdurgin) [inline src/osd/osd_types.h]: what's the benefit of using string_view when we always convert these to string for storing in a container? is the temporary somehow elided in the encode() case?

**Comment by jdurgin** [comment]
' osd/: Move PG::_prepare_write_info into osd_types prepare_info_keymap ' has your email in the commit message twice instead of your name, otherwise everything looks good. Much nicer to have RAII!

File: src/osd/osd_types.h (hunk 1 above)
Comment (athanatos) [reply src/osd/osd_types.h]: Crimson uses string_view a lot internally.  I think it's a wash for classic, but it's convenient for crimson.

File: src/crimson/osd/osd.cc [hunk 2]
```diff
@@ -238,7 +238,7 @@ seastar::future<> OSD::_preboot(version_t oldest, version_t newest)
     logger().warn("osdmap NOUP flag is set, waiting for it to clear");
   } else if (!osdmap->test_flag(CEPH_OSDMAP_SORTBITWISE)) {
     logger().error("osdmap SORTBITWISE OSDMap flag is NOT set; please set it");
-  } else if (osdmap->require_osd_release < CEPH_RELEASE_LUMINOUS) {
+  } else if (osdmap->require_osd_release < ceph_release_t::luminous) {
```
Comment (tchaikov) [inline src/crimson/osd/osd.cc]: @athanatos Sam could you drop this change from your PR? i just realized that it dups with #28053.

Comment (athanatos) [reply src/crimson/osd/osd.cc]: Sure, will do.

**Pull request merged**